services/src/main/java/org/keycloak/authentication/requiredactions/RecoveryAuthnCodesAction.java

```diff
@@ -98,6 +99,37 @@ public void processAction(RequiredActionContext reqActionContext) {
         generatedAtTime = Long.parseLong(httpReqParamsMap.getFirst(FIELD_GENERATED_AT_HIDDEN));
         generatedUserLabel = httpReqParamsMap.getFirst(FIELD_USER_LABEL_HIDDEN);
 
+        // Retrieve the initially generated recovery codes from the authentication session
+        String storedCodes = reqActionContext.getAuthenticationSession().getAuthNote(AUTH_NOTE_INITIAL_GENERATED_CODES);
+        if (storedCodes == null) {
+            Response challenge = reqActionContext.form()
+                    .setError("No initial recovery codes found.")
+                    .createErrorPage(Response.Status.BAD_REQUEST);
+            reqActionContext.challenge(challenge);
+            return;
+        }
+        List<String> initialGeneratedCodes = Arrays.asList(storedCodes.split(","));
+
+        // Check the number of recovery codes
+        if (generatedCodes.size() != initialGeneratedCodes.size()) {
+            Response challenge = reqActionContext.form()
+                    .setError("Invalid number of recovery codes.")
+                    .createErrorPage(Response.Status.BAD_REQUEST);
+            reqActionContext.challenge(challenge);
+            return;
+        }
+
+        // Check the content of the recovery codes
+        for (String code : generatedCodes) {
+            if (!initialGeneratedCodes.contains(code)) {
+                Response challenge = reqActionContext.form()
+                        .setError("Invalid recovery code.")
+                        .createErrorPage(Response.Status.BAD_REQUEST);
+                reqActionContext.challenge(challenge);
+                return;
+            }
+        }
```

Comment:
Rather than checking these manually, perhaps we can use the `equals()` method of the list to check these. They should always be in the same order AFAIK.


```java
if (!generatedCodes.equals(initialGeneratedCodes)) {
    Response challenge = reqActionContext.form()
            .setError("Submitted recovery codes do not match generated ones.")
            .createErrorPage(Response.Status.BAD_REQUEST);
    reqActionContext.challenge(challenge);
    return;
}
````